pkg/controller/persistentvolume/persistentvolume_claim_binder_controller.go

```diff
@@ -282,6 +282,15 @@ func syncClaim(volumeIndex *persistentVolumeOrderedIndex, binderClient binderCli
 			return fmt.Errorf("A volume match does not exist for persistent claim: %s", claim.Name)
 		}
 
+		// create a reference to the claim and assign it to the volume being bound.
+		// the volume is a pointer and assigning the reference fixes a race condition where another
+		// claim might match this volume but before the claimRef is persistent in the next case statement
+		claimRef, err := api.GetReference(claim)
+		if err != nil {
+			return fmt.Errorf("Unexpected error getting claim reference: %v\n", err)
+		}
+		volume.Spec.ClaimRef = claimRef
```

Comment:
Yes, you're right about unsetting in the event of error.  I will add that.

As for your first comment, there is a map of volumes in memory (the local store).  By manipulating the pointer, it won't be bound to another PVC.

This, however, does not survive cluster crash and re-opens the race condition until the PVC can be updated again and its claimRef persistent to the backing PV.  For this reason, I believe the bind is in the wrong place (on pvc.Spec.VolumeName instead of pv.Spec.ClaimRef).